src/linear_algebra/matrix/transvection.lean

```diff
@@ -0,0 +1,754 @@
+  transvections, and invariant under product, is true for all matrices.
+* `diagonal_transvection_induction_of_det_ne_zero` is the same statement over invertible matrices.
+
+## Implementation details
+
+The proof of the reduction results is done inductively on the size of the matrices, reducing an
+`(r + 1) × (r + 1)` matrix to a matrix whose last row and column are zeroes, except possibly for
+the last diagonal entry. This step is done as follows.
+
+If all the coefficients on the last row and column are zero, there is nothing to do. Otherwise,
+one can put a nonzero coefficient in the last diagonal entry by a row or column operation, and then
+subtract this last diagonal entry from the other entries in the last row and column to make them
+vanish.
+
+This step is done in the type `fin r ⊕ unit`, where `fin r` is useful to choose arbitrarily some
+order in which we cancel the coefficients, and the sum structure is useful to use the formalism of
+block matrices.
+
+To proceed with the induction, we reindex our matrices to reduce to the above situation.
+-/
+
+universes u₁ u₂
+
+namespace matrix
+open_locale matrix
+
+variables (n p : Type*) (R : Type u₂) {𝕜 : Type*} [field 𝕜]
+variables [decidable_eq n] [decidable_eq p]
+variables [comm_ring R]
+
+section elementary_basis
+
+variables {n} (i j : n)
+
+/-- The elementary matrix `E i j` has a one at position `(i, j)`, and zeroes elsewhere.
+It is useful to define these matrices for explicit calculations in sl n R. -/
+@[reducible] definition E : matrix n n R := λ i' j', if i = i' ∧ j = j' then 1 else 0
+
+@[simp] lemma E_apply_one : E R i j i j = 1 := if_pos (and.intro rfl rfl)
+
+@[simp] lemma E_apply_zero (i' j' : n) (h : ¬(i = i' ∧ j = j')) : E R i j i' j' = 0 := if_neg h
+
+@[simp] lemma E_diag_zero (h : j ≠ i) : matrix.diag n R R (E R i j) = 0 :=
+funext $ λ k, if_neg $ λ ⟨e₁, e₂⟩, h (e₂.trans e₁.symm)
+
+variable [fintype n]
+
+lemma E_trace_zero (h : j ≠ i) : matrix.trace n R R (E R i j) = 0 := by simp [h]
+
+@[simp] lemma E_mul_apply (b : n) (M : matrix n n R) :
+  (E R i j ⬝ M) i b = M j b :=
+by simp [matrix.mul_apply]
+
+@[simp] lemma mul_E_apply (a : n) (M : matrix n n R) :
+  (M ⬝ E R i j) a j = M a i :=
+by simp [matrix.mul_apply]
+
+@[simp] lemma E_mul_apply_of_ne (a b : n) (h : a ≠ i) (M : matrix n n R) :
+  (E R i j ⬝ M) a b = 0 :=
+by simp [matrix.mul_apply, h.symm]
+
+@[simp] lemma mul_E_apply_of_ne (a b : n) (hbj : b ≠ j) (M : matrix n n R) :
+  (M ⬝ E R i j) a b = 0 :=
+by simp [matrix.mul_apply, hbj.symm]
+
+@[simp] lemma E_mul_E (k : n) : E R i j ⬝ E R j k = E R i k :=
+begin
+  ext a b,
+  simp only [matrix.mul_apply, boole_mul],
+  by_cases h₁ : i = a; by_cases h₂ : k = b;
+  simp [h₁, h₂],
+end
+
+@[simp] lemma E_mul_E_of_ne {k l : n} (h : j ≠ k) : E R i j ⬝ E R k l = 0 :=
+begin
+  ext a b,
+  simp only [matrix.mul_apply, dmatrix.zero_apply, boole_mul],
+  by_cases h₁ : i = a;
+  simp [h₁, h, h.symm],
+end
+
+end elementary_basis
```

Comment:
Are you ok with waiting for #9088, then putting all of this at the end of `data/matrix/basis.lean`? Actually deduplicating it with `std_basis_matrix` can be a TODO comment, but if it's in the same file it will at least be obvious that the duplication is there.